### PR TITLE
[Rewriter] Support struct type used by array.

### PR DESCRIPTION
**Pull request by python3kgae**
Save element type of array type when collect visited type.

Fixes 5255 https://github.com/microsoft/DirectXShaderCompiler/issues/5255

### CODE REVIEW

File: tools/clang/test/HLSLFileCheck/rewriter/struct_array.hlsl [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// RUN: %dxr -remove-unused-functions -remove-unused-globals -E vs_main %s | FileCheck %s
+
+// Make sure InstanceDataStructType not removed.
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/rewriter/struct_array.hlsl:3]: Fixed.

File: tools/clang/tools/libclang/dxcrewriteunused.cpp [hunk 2]
```diff
@@ -301,6 +305,10 @@ class TypeVisitor : public RecursiveASTVisitor<TypeVisitor> {
     }
     return true;
   }
+  bool VisitArrayType(const ArrayType *AT) {
+    TraverseType(AT->getElementType());
+    return true;
+  }
```
Comment (pow2clk) [inline tools/clang/tools/libclang/dxcrewriteunused.cpp:311]: The corresponding DataRecursiveASTVistor code is a bit hard to follow due to macro skullduggery, but I don't think this is ever called. The only use of TypeVisitor is https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/tools/libclang/dxcrewriteunused.cpp#L1462
The implementation of which is here
https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/include/clang/AST/DataRecursiveASTVisitor.h#L548
Which defines ABSTRACT_TYPE as nothing and for Arrays, this is the macro that Array uses:
https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/include/clang/AST/TypeNodes.def#L70

So I really don't see how this could be called. Am I missing something?

**Review by pow2clk — COMMENTED** [review]
I have one major uncertainty about VisitArrayType, but the description could really be more informative. You might mention how the SaveTypeDecl is used for appropriate tag types, but how it previously ignored array types that contained tag types and how this passes through that information.

If the VisitArrayType function is not dead as I have concluded that it is, some description of how that works would be helpful too.

File: tools/clang/tools/libclang/dxcrewriteunused.cpp (hunk 2 above)
Comment (python3kgae) [reply tools/clang/tools/libclang/dxcrewriteunused.cpp]: VisitArrayType will be called here https://github.com/microsoft/DirectXShaderCompiler/blob/c2233c6c0a528ecaf5d2d008684e08daf722d845/tools/clang/include/clang/AST/DataRecursiveASTVisitor.h#L350

But the change is still useless because VisitRecordType will be invoked anyway with/without the change. :)

Removed the VisitArrayType.

**Comment by damyanp** [comment]
This PR was closed as it has not been updated in the last two years. Please feel free to reopen if this PR should be merged and is in a reviewable state.

**Comment by damyanp** [comment]
This PR was closed as it has not been updated in the last two years. Please feel free to reopen if this PR should be merged and is in a reviewable state.
